### ISSUE TITLE
Actions prevented by RLS not communicated in GraphQL response (401 Access Denied or error object)

### ISSUE THREAD

**Issue by arvalaan** [supabase/pg_graphql#186]
**Describe the bug**
When attempting to update a row on a table with RLS enabled, where the policy forbids the user to do so, no error is thrown whatsoever.

**To Reproduce**
1. Create Table `updates`
2. Configure columns id (default: uuid), title, content, created_by (default: uid())
3. Enable RLS
4. Create Policy for UPDATE
```
CREATE POLICY "Enable update for users based on user_id" ON "public"."updates"
AS PERMISSIVE FOR UPDATE
TO public
USING (auth.uid() = user_id)
```
5. Send mutation request as authenticated user
```
mutation {
	updateupdatesCollection(
		set: { title: "Updated Title" }
		filter: { id: { eq: 22 } }
	) {
		records {
			id
			title
			content
		}
	}
}
```
6. Observe response

**Expected behavior**
Error thrown in response that user fails RLS

**Additional context**
This is my first time working with GraphQL, it might be me who is wrong. I couldn't find anything in the Docs about how to handle these situations where I want to make the user aware that their update has not been processed.

A workaround I thought of was counting the amount of objects I'm setting and checking that vs the `affectedCount`

**Comment by olirice** [comment]
The reason no error is returned is because RLS filters out the record being updated

The equivalent SQL
```sql
update my_table set title = 'Updated Title' where id = 22;
```
would also execute without error with 0 records being impacted

---

It would be possible to make this return an error since `select` permissions are a pre-req for `update` and `delete` but I think returning an `affectedCount` of 0 might be "more correct" than returning an error.

Its clear from the context of your query that an error would be more useful in this case, but there may be other situations (bulk updates) where its less obvious what the user's goal is

---

@steve-chavez do you know if PostgREST treats this situation as an error or silently refuse to do the update? Do you think that is the "correct" solution?

**Comment by arvalaan** [comment]
@olirice Am I forced to handle this client side or can I do something in Supabase with the expected updates and the affectedCount?

**Comment by olirice** [comment]
Yes, this would need to be handled client side by referencing the `affectedCount`

**Comment by steve-chavez** [comment]
>@steve-chavez do you know if PostgREST treats this situation as an error or silently refuse to do the update?

The same as the default pg behaviour, violating the USING condition doesn't throw an error. Violating the WITH CHECK condition does throw an error. 

> Do you think that is the "correct" solution?

Hm, usually this is handled by using the [.single()](https://supabase.com/docs/reference/javascript/single) method in the postgrest-js side. This enforces a `rowcount`(affected rows for this case) of 1, if not complied it throws an error.

It's not ideal. We have a pending enhancement that would allow us to control the `rowcount` arbitrarily through a header(`Prefer: rowcount=<lte/gt/eq>.1234`). Perhaps a global `rowcount` can be set to turn the USING failure to an error, but I'm not clear on the design yet.

**Comment by olirice** [comment]
thanks!

---

@arvalaan applying that to your case, you could force an error message by

- update the `using` clause to allow visibility during updates to match `selects`
- add a `with check` clause to validate `auth.uid() = user_id`
- revoke `update` on `public.updates.user_id` so it can't be edited after insert

Ex:
```sql
CREATE POLICY "Enable update for users based on user_id" ON "public"."updates"
AS PERMISSIVE FOR UPDATE
TO public
USING (true) -- this should match your select policy
WITH CHECK (auth.uid() = user_id) 
```

and also restricting the the role's column level permissions to prevent the `user_id` column from being updated

```sql
revoke update (user_id) on "public"."updates" from authenticated, anon;
```
